### PR TITLE
Speed-up adaptive average pooling for the common case of size=1 output

### PR DESCRIPTION
**Pull request by ngimel**
When adaptive pooling has to produce a single pixel feature map, it is faster to do so by calling .mean(). Backward calls a pretty inefficient cuda kernel with atomics, which becomes ridiculously slow for halfs. For half this PR provides approx 30x speed-up for adaptive average pooling, which results in 30% end-to-end speed-up on senet. Improvements are smaller for float, but still significant (approx 5x). 
Also this PR unifies handling of 3d (no batch dimension) and 4d tensors, using negative dimension indices. 
cc @ezyang for review.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Review by soumith — APPROVED** [review]
CI failures are unrelated

**Comment by ezyang** [comment]
This breaks the XLA build:

```
Feb 15 05:39:04 In file included from torch_xla/csrc/aten_xla_type.cpp:1:
Feb 15 05:39:04 /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.h:192:68: error: only virtual member functions can be marked 'override'
Feb 15 05:39:04       const at::Tensor& grad_output, const at::Tensor& self) const override;
Feb 15 05:39:04                                                                    ^~~~~~~~
Feb 15 05:39:08 torch_xla/csrc/aten_xla_type.cpp:577:12: error: no member named 'adaptive_avg_pool2d_backward' in 'torch_xla::AtenXlaTypeBase'; did you mean simply 'adaptive_avg_pool2d_backward'?
Feb 15 05:39:08     return AtenXlaTypeBase::adaptive_avg_pool2d_backward(grad_output, self);
Feb 15 05:39:08            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Feb 15 05:39:08            adaptive_avg_pool2d_backward
Feb 15 05:39:08 torch_xla/csrc/aten_xla_type.cpp:574:25: note: 'adaptive_avg_pool2d_backward' declared here
Feb 15 05:39:08 at::Tensor AtenXlaType::adaptive_avg_pool2d_backward(
Feb 15 05:39:08                         ^
Feb 15 05:39:08 torch_xla/csrc/aten_xla_type.cpp:582:12: error: no member named 'adaptive_avg_pool2d_backward' in 'torch_xla::AtenXlaTypeBase'; did you mean simply 'adaptive_avg_pool2d_backward'?
Feb 15 05:39:08     return AtenXlaTypeBase::adaptive_avg_pool2d_backward(grad_output, self);
Feb 15 05:39:08            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Feb 15 05:39:08            adaptive_avg_pool2d_backward
Feb 15 05:39:08 torch_xla/csrc/aten_xla_type.cpp:574:25: note: 'adaptive_avg_pool2d_backward' declared here
Feb 15 05:39:08 at::Tensor AtenXlaType::adaptive_avg_pool2d_backward(
Feb 15 05:39:08                         ^
Feb 15 05:39:10 3 errors generated.
```

cc @ailzhang

**Comment by ngimel** [comment]
Now `at::adaptive_avg_pool2d` follows a pattern of other functions like `at::conv`, `at::batch_norm`, `at::layer_norm`, `at::softmax` etc, where top level function does not have a defined derivative, and rather is a composition of differentiable functions, with exact path determined at runtime. Thus, there is no longer `at:adaptive_avg_pool2d_backward`, but rather `at::_adaptive_avg_pool2d_backward`, to reflect the fact that it's no longer top level backward function. I can return it to the way it was at the expense of some ugliness and code repetition, but given that other functions are already using this pattern I think the proper fix is on torch_xla side. I'm happy to look into this if someone gives me some pointers.

**Comment by ezyang** [comment]
Yes, I think the correct fix is to just adjust the XLA definition to use the new signature.
